JumpscaleLibs/sal/zosv2/resource.py

```diff
@@ -72,16 +78,56 @@ def payout_farmers(self, resource_units_per_node, reservation_id):
         tfchain_wallet = tfchain_client.wallets.get("main")
         transactions = []
         for resource_unit_node in resource_units_per_node:
+            resource_unit_node.workload_id.sort()
+            workload_ids = [str(wid) for wid in resource_unit_node.workload_id()]
+            msg = "-".join(workloads_ids)
+            msg = "{}-{}".format(reservation_id, msg)
             (txn, submitted) = tfchain_wallet.coins_send(
-                recipient=resource_unit_node.farm_wallet,
-                amount=str(resource_unit_node.TOTAL_COST),
-                data=f"{resource_unit_node.workload_id}-{reservation_id}",
+                recipient=resource_unit_node.farm_wallet, amount=str(resource_unit_node.TOTAL_COST), data=msg,
             )
             if submitted:
                 transactions.append(txn)
                 continue
         return transactions
 
+    def validate_reservation_payment(self, reservation_id):
+        me_tid = j.tools.threebot.me.default.tid
+        # load all farms belonging to our threebot id
+        farms = self._actor_directory.farms.owned_by(me_tid).farms
+        farm_ids = set()
+        for farm in farms:
+            farm_ids.add(farm.id)
+        # load reservation and parse resource units
+        resource_units_per_node = self.calculate_used_resources_cost()
+        # tuple of the message and amount expected in a tx
+        expected_txes = set()
+        for resource_unit_node in resource_units_per_node:
+            if resource_unit_node.farm_id in farm_ids:
+                resource_unit_node.workload_id.sort()
+                workload_ids = [str(wid) for wid in resource_unit_node.workload_id()]
+                msg = "-".join(workloads_ids)
+                msg = "{}-{}".format(reservation_id, msg)
+                amount = str(resource_unit_node.TOTAL_COST)
+                expected_txes.add((msg, amount))
+        # load the transactions in our wallet, for every workload in one of the owned farms, whe need a trandaction
+        # with the right amount of tft
+        tfchain_client = j.clients.tfchain.get("tfchain")
```

Comment:
This is not ok. You cannot expect user to always have a wallet called tfchain.
The wallet must be an argument of this method.